flatbuffers/flatbuffers_case_study.md

```diff
@@ -0,0 +1,151 @@
+EventSerializer.Event.addDataType(builder, EventSerializer.EventData.MousePositionData)
+EventSerializer.Event.addData(builder, mousePositionDataOffset)
+const eventOffset = EventSerializer.Event.endEvent(builder)
+
+EventSerializer.Event.finishEventBuffer(builder, eventOffset)
+
+const bytes = builder.asUint8Array() // our data bytes
+```
+
+These bytes are stored for later use. We simply store this data in Amazon S3.
+
+Reading the bytes looks like this:
+
+```javascript
+const buffer = new flatbuffers.ByteBuffer(bytes)
+const event = EventSerializer.Event.getRootAsEvent(buffer)
+
+console.log('event timestamp', event.timestamp())
+console.log('event type', event.type())
+console.log('event data type', event.dataType())
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
+
+# Phases of refactor
+Refactoring our entire flow from JSON to FlatBuffers is quite a feat. It's a architecture crossing change that will be dangerous to try and accomplish in one go. We've considered this issue for a while and managed to split the redactor into several independent steps.
+
+## FB in recorder
+We began in transforming our recorder messages to FlatBuffers. The recorder creates and send the recorded events as FlatBuffers. The first step of the processing pipeline is to transform the binary data back to JSON allowing the rest of the architecture to stay unchanged.
+
+There is good value to this seemingly small change. It would allow us to skip the process blocking JSON.stringify, and minimize the amount of data the end user needs to upload to our server.
+
+## FB in player
+The second step we took was to add FlatBuffers support to our player. Naturally we needed to convert the JSON data we store to FlatBuffers but this extra converting code is a small price to pay in order to achieve a factorize a big refactor.
+
+## Store FB - "point of no return"
+The Final step is to actually store the FlatBuffers bytes in our storage. This is the "point of no return". A benefit of the previous steps is that we could easily roll back the change. Once we loose the JSON all together a rollback would be very difficult.
+
+In fact we have not taken this step yet. A Major benefit of JSON is that it's easily streamable using libraries such as [Oboe.js](http://oboejs.com/). We have yet to discover a method to achieve a similar goal using FlatBuffers.
+
+# benchmarks
+We have expected FlatBuffers to reduce the size of our data. We've converted some of our JSON sessions to FlatBuffers to check this hypothesis. Each tested session
+
+<scatter plot>
+
+The results where quite impressive! 40% - 94% size decrease in our test data.
+
+# Drawbacks
+The advantages FlatBuffers offers do not come for free.
+
+* Non human readable
+* Verbal and cumbersome API
+* Backwards compatibility issues
+* Lack of streaming support
```

Comment:
I would write a conclusion
And also touch put in the benchmarks speed comparison.